### ISSUE TITLE
Code freezes when it receives focus

### ISSUE THREAD

**Issue by skaldesh** [microsoft/vscode#154509]
Issue Type: <b>Performance Issue</b>

The issue itself is very simple to explain: When VSCode is already up and running and I switch the focus to anonther application (e.g. Browser, File Explorer, ...) and then **after some time (>~5s)** switch focus back to VSCode, the window just freezes for *1-3s* before everything is back to normal.  
Has happened to me consistently over the last 6 months at least, happened in each and every project.  
It is something that is unfortunately not reliable to reproduce. Sometimes I can code a few hours or days, before it hits me, but sometimes it happens very frequent as well.  

Only solution right now is to completely restart all VSCode instances.

I did a performance recording as suggested [here](https://github.com/microsoft/vscode/wiki/Performance-Issues#visual-studio-code-is-sluggish), and uploaded it to this issue here: [vscode-pprof.json.txt](https://github.com/microsoft/vscode/files/9072436/vscode-pprof.json.txt)

**The bug is not related to extensions, since I reproduced it while all extensions were disabled**

VS Code version: Code - OSS 1.68.1 (30d9c6cd9483b2cc586687151bcbcd635f373630, 2022-06-16T08:22:04.269Z)
OS version: Linux x64 5.15.50-1-lts
Restricted Mode: No

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|AMD Ryzen 7 4800U with Radeon Graphics (16 x 1800)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: disabled_off<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>oop_rasterization: disabled_off<br>opengl: enabled_on<br>rasterization: disabled_software<br>raw_draw: disabled_off_ok<br>skia_renderer: enabled_on<br>video_decode: disabled_software<br>video_encode: disabled_software<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|3, 2, 2|
|Memory (System)|15.00GB (1.28GB free)|
|Process Argv|--unity-launch|
|Screen Reader|no|
|VM|0%|
|DESKTOP_SESSION|undefined|
|XDG_CURRENT_DESKTOP|undefined|
|XDG_SESSION_DESKTOP|undefined|
|XDG_SESSION_TYPE|wayland|
</details><details>
<summary>Process Info</summary>

```
CPU %	Mem MB	   PID	Process
    0	   200	  5727	code-oss main
    0	    77	  5730	   zygote
    0	   200	  5756	     gpu-process
    0	    77	  5731	   zygote
    0	     0	  5733	     zygote
    0	    31	 37896	       utility
    0	   108	 38912	       issue-reporter
    0	    92	  5791	   utility-network-service
    0	   261	  5804	   window (backend.go - vision - Code - OSS)
    3	   277	  5805	   window (daemon.go - nline - Code - OSS)
    0	   277	  5811	   window (semanticsegmentationmodel_test.go - nlib - Code - OSS)
    0	   169	  5896	   shared-process
    0	   108	  5975	     ptyHost
    0	     0	  7243	       /usr/bin/bash
    0	     0	  7262	       /usr/bin/bash
    0	     0	  7333	       /usr/bin/bash
    0	     0	 14638	       /usr/bin/bash
    0	    92	  6016	     fileWatcher
    0	    92	  6017	     fileWatcher
    0	    92	 14598	     fileWatcher
    0	    92	 38171	     fileWatcher
    0	     0	 38930	     /usr/bin/ps -ax -o pid=,ppid=,pcpu=,pmem=,command=
    0	   154	  5897	   extensionHost
    0	   184	  6146	     /home/skaldesh/go/bin/gopls -mode=stdio
    0	   154	  5934	   extensionHost
    0	   246	  6420	     /home/skaldesh/go/bin/gopls -mode=stdio
    0	   292	 14514	   window (camera.go - camera - Code - OSS)
    0	   154	 14586	   extensionHost
    0	   230	 14684	     /home/skaldesh/go/bin/gopls -mode=stdio
    1	   138	 38158	   extensionHost
    0	   277	 38435	     /home/skaldesh/go/bin/gopls -mode=stdio
```

</details>
<details>
<summary>Workspace Info</summary>

```
|  Window (camera.go - camera - Code - OSS)
|  Window (semanticsegmentationmodel_test.go - nlib - Code - OSS)
|  Window (daemon.go - nline - Code - OSS)
|  Window (backend.go - vision - Code - OSS)
|    Folder (camera): 811 files
|      File types: go(79) h(24) cpp(21) yaml(4) md(2) gitignore(1) pprof(1)
|                  tar(1) txt(1) mod(1)
|      Conf files:
|    Folder (nline): 1520 files
|      File types: go(64) xml(8) md(2) npack(2) yaml(2) gitignore(1) iml(1)
|                  json(1) tar(1) txt(1)
|      Conf files: settings.json(1) dockerfile(1)
|    Folder (nlib): 6451 files
|      File types: jpg(5030) go(82) yaml(41) npack(30) h(25) cpp(22) xml(12)
|                  onnx(12) md(9) pt(9)
|      Conf files: launch.json(1)
|      Launch Configs: go
|    Folder (vision): 3619 files
|      File types: qml(578) png(182) qmlc(170) dll(106) o(51) qmltypes(39)
|                  cpp(24) h(18) js(13) go(12)
|      Conf files: makefile(3);
```

</details>
<details><summary>Extensions (8)</summary>

Extension|Author (truncated)|Version
---|---|---
systemd-unit-file|coo|1.0.6
dart-code|Dar|3.44.0
go|gol|0.34.1
vscode-simple-icons|Lau|1.16.0
hexeditor|ms-|1.9.7
orbit|r0l|0.0.1
cmake|twx|0.0.17
vscode-todo-highlight|way|1.0.5

</details>

**Comment by VSCodeTriageBot** [comment]
Thanks for creating this issue! It looks like you may be using an old version of VS Code, the latest stable release is 1.69.0. Please try upgrading to the latest version and checking whether this issue remains.

Happy Coding!

**Comment by r0l1** [comment]
Having the same issue...

**Comment by lorand-horvath** [comment]
You have to clear the VSCode cache, as I described here https://github.com/microsoft/vscode/issues/146737#issuecomment-1153128058
This is caused by large amounts of cached files compounding and giving sluggish performance after some time. The only solution I have found was to completely uninstall, remove the cache folders and reinstall VSCode.
You already know all of this, but unfortunately we cannot continue commenting in that issue https://github.com/microsoft/vscode/issues/146737 because for some reason someone locked it up. 

Anyway, you should have mentioned that issue here and also my response, even if it annoys you having to clean the cache. You want to find another, easier solution, without having to reinstall VSCode, I get it. But believe me, you won't find any. I've been pulling out my hair for many months until I nuked everything and started fresh.

For further info, I'll post my response from https://github.com/microsoft/vscode/issues/146737 here as well, since that issue is locked:
-------
The solution to get rid of the slowness and get back the snappiness and quick reaction when focusing back to VSCode from another program is to uninstall VSCode, restart the PC and completely remove the folder `C:\Users\[User]\AppData\Roaming\Code`, which in my case had 380 MB and several thousand files cached in it. The folder has grown so much over 2 years of daily VSCode use, with regular upgrading to newest version as soon as it was released. I've also removed the VSCode extensions folder `C:\Users\[User]\.vscode`, I'm not sure whether that helped or not.

After reinstalling VSCode I had to reconfigure everything from scratch and install all extensions, but the cache folder is now only 80 MB and contains 550 files. Working with VSCode is a breeze again...

*Credit goes to @icedterminal https://github.com/microsoft/vscode/issues/146737#issuecomment-1124494187 - thanks for the idea!

*Cross-referencing possibly related issues, perhaps it helps someone else not noticing this solution: https://github.com/microsoft/vscode/issues/123257 https://github.com/microsoft/vscode/issues/141735 https://github.com/microsoft/vscode/issues/147151 https://github.com/microsoft/vscode/issues/147188

Edit: Slow file open fixed as well - brilliant!

**Comment by tanhakabir** [comment]
@lorand-horvath thank you so much for your help! Yes this looks like a duplicate of #146737 and an upstream issue. 

Going to close this issue as a duplicate for now but @skaldesh let me know if @lorand-horvath's suggestions don't help you

**Comment by icedterminal** [comment]
> You want to find another, easier solution, without having to reinstall VSCode, I get it. But believe me, you won't find any.

There actually is an easier workaround. The program files themselves aren't an issue in any capacity so there is no need to reinstall Code. Deleting the entirety of `%AppData%\Code` means you lose all workspaces you have to configure again and again. What I tried was more of a drastic "I need to find a solution now" trial and error. While it did work, https://github.com/microsoft/vscode/issues/146737#issuecomment-1153131501 made excellent point. I completely forgot to revisit this topic but since I have been motioned here, I'll leave this bit of information.

Close Code and simply delete the following directories:

![image](https://user-images.githubusercontent.com/4755734/178086335-2f70c1e5-d490-4306-9361-9ed0a244c20a.png)

Code lacks a method to clear all caches. The user must manually do so to restore expected performance.

**Comment by lorand-horvath** [comment]
@icedterminal Removing only the _cache_ folders makes a lot of sense and this should probably be officially implemented in VSCode. Thanks!

**Comment by r0l1** [comment]
@tanhakabir this is VSCode issue on Linux. Please reopen. Deleting the cache in `~/config/Code/Cache*` or `~/config/Code OSS/Cache*` didn't seam to fix the issue.

**Comment by lorand-horvath** [comment]
@skaldesh Can you please check how many files do you have in the folder `C:\Users\[User]\AppData\Roaming\Code`?

**Comment by skaldesh** [comment]
I am on Linux

**Comment by skaldesh** [comment]
I am on Linux, but my `.config/Code - OSS` folder is 410MB large.

**Comment by skaldesh** [comment]
```
52K	.config/Code - OSS/Backups
8.0K	.config/Code - OSS/blob_storage
20K	.config/Code - OSS/Cache
16K	.config/Code - OSS/CachedConfigurations
6.6M	.config/Code - OSS/CachedData
900K	.config/Code - OSS/CachedExtensions
36K	.config/Code - OSS/Code Cache
4.0K	.config/Code - OSS/code.lock
28K	.config/Code - OSS/Cookies
0	.config/Code - OSS/Cookies-journal
8.0K	.config/Code - OSS/Crashpad
1.7M	.config/Code - OSS/Crash Reports
32K	.config/Code - OSS/databases
448K	.config/Code - OSS/Dictionaries
8.0K	.config/Code - OSS/exthost Crash Reports
8.0K	.config/Code - OSS/fileWatcher
280K	.config/Code - OSS/GPUCache
4.0K	.config/Code - OSS/languagepacks.json
28K	.config/Code - OSS/Local Storage
2.8M	.config/Code - OSS/logs
4.0K	.config/Code - OSS/machineid
4.0K	.config/Code - OSS/Network Persistent State
3.3M	.config/Code - OSS/Preferences
452K	.config/Code - OSS/QuotaManager
0	.config/Code - OSS/QuotaManager-journal
4.0K	.config/Code - OSS/rapid_render.json
348M	.config/Code - OSS/Service Worker
36K	.config/Code - OSS/Session Storage
4.0K	.config/Code - OSS/TransportSecurity
46M	.config/Code - OSS/User
4.0K	.config/Code - OSS/Workspaces
```
Most of that is in ServiceWorker

**Comment by lorand-horvath** [comment]
On Windows I had 380 MB in about 20k files, which is insane, hence the slowness. In a fresh/clean install there should be less than 1k files, normally a few hundred only. Can you check how many files do you have in there? You might have to uninstall VSCode and completely remove that folder (if there are too many files). I couldn't tell you the number above which the slowness occurs, but I would make a guesstimate somewhere around 5k (also depending on your hardware, especially the CPU).

**Comment by skaldesh** [comment]
Accumulated number of files in each directory.
```
./databases:	2
./Backups:	2
./Crashpad:	1
./Code Cache:	4
./Local Storage:	6
./blob_storage:	0
./Session Storage:	6
./Crash Reports:	4
./Workspaces:	0
./CachedConfigurations:	1
./fileWatcher:	1
./GPUCache:	5
./logs:	477
./CachedData:	98
./User:	4681
./CachedExtensions:	2
./exthost Crash Reports:	1
./Cache:	2
./Service Worker:	9974
./Dictionaries:	1
```

**Comment by lorand-horvath** [comment]
There you have it: `Service Worker` directory has almost 10000 files... I only have 22 files in it. I think there was a bug in some older VSCode version which for some reason filled up that directory.

**Comment by skaldesh** [comment]
Inside `Service Worker`, I have 3 directories with the following number of files:  
```
.config/Code - OSS/Service Worker/Database:	10
.config/Code - OSS/Service Worker/CacheStorage:	7760
.config/Code - OSS/Service Worker/ScriptCache:	2204
```

Do you think it is safe to just delete `CacheStorage` and `ScriptCache`?

**Comment by skaldesh** [comment]
Unfortunately, it did not solve the issue.   
My directory now looks like this:  
```
24K	Backups
8.0K	blob_storage
20K	Cache
16K	CachedConfigurations
14M	CachedData
880K	CachedExtensions
36K	Code Cache
4.0K	code.lock
28K	Cookies
0	Cookies-journal
8.0K	Crashpad
1.7M	Crash Reports
32K	databases
448K	Dictionaries
8.0K	exthost Crash Reports
8.0K	fileWatcher
280K	GPUCache
4.0K	languagepacks.json
28K	Local Storage
2.8M	logs
4.0K	machineid
4.0K	Network Persistent State
3.3M	Preferences
452K	QuotaManager
0	QuotaManager-journal
4.0K	rapid_render.json
640K	Service Worker
44K	Session Storage
4.0K	TransportSecurity
46M	User
4.0K	Workspaces
```

In User, we have this:  
```
1.8M	User/globalStorage
29M	User/History
4.0K	User/keybindings.json
4.0K	User/settings.json
4.0K	User/snippets
16M	User/workspaceStorage
```

Maybe it is a just a bug in VSCode after all, and not some caching issue.  
I was able to fairly quickly reproduce the bug this time. I opened three workspaces in quick succession and got the freeze immediately when switching focus between their windows

**Comment by skaldesh** [comment]
### **Important Update**  
I think I just discovered maybe the steps to reproduce this.  
My current setup is that I have 3 VSCode windows open with one workspace each, let's call them A, B and C.  
When I have the focus on the editor in **all three windows**, I can switch smoothly between them.  

However, when I have the **integrated terminal** open in e.g. A, then VSCode freezes when switching from A to B, if the focus was in A's terminal. If the focus was on the editor, no freeze.  

I can reproduce the freeze right now with this setup almost always. There are a few occasions, where it does not freeze (~ 1 out of 10).

**Comment by lorand-horvath** [comment]
@skaldesh Great find! tagging @deepak1556 and @Tyriar

**Comment by skaldesh** [comment]
I have to say, though, that I just got the freeze also once when switching between the editors only.  
But maybe my finding above helps to more consistently reproduce this. My colleague experiences these freezes as well, but he does not use the integrated terminal at all, as far as I am aware.  

So take it with a grain of salt. :)

**Comment by skaldesh** [comment]
Is there any update here? If more profiling etc. is needed I am happy to help

**Comment by lorand-horvath** [comment]
@skaldesh As I said above, the issue has been completely solved in my case only after doing a clean installation (removed everything left behind after the uninstall). Have you tried doing that?

**Comment by satgo1546** [comment]
Same issue here on Windows 11. Deleting the cache in `%APPDATA%\Code\Cache*` (~200MB) didn't help, sadly.  I can confirm that a complete reinstall after removing `%APPDATA%\Code` and `%USERPROFILE%\.vscode` indeed resolves the issue for me.

I suspect the native window occlusion calculation must have something to do with the freeze in my case, since there's no lag if the VSCode window stays visible on screen while the focus switches back and forth. Occlusion calculation itself doesn't seem to be the culprit, though, but it might be related to the fact that it is unreliable to reproduce the freeze.

**Comment by skaldesh** [comment]
@satgo1546 Interesting, I need to check this for myself if having the two windows both visible does not lead to a freeze. Will update here

**Comment by skaldesh** [comment]
I can confirm the behaviour on Linux that @satgo1546 observed!  
Once I got the freeze again, I moved the windows next to each other on the same screen (both visible at the same time).  
The freeze then did not occur anymore

**Comment by skaldesh** [comment]
I have reinstalled vscode completetly from scratch and only copied my "settings.json" and "keybindings.json" files.  
After 1 day, the freeze is happening again.

**Comment by r0l1** [comment]
I can confirm this._

**Comment by skaldesh** [comment]
@r0l1  and I realized we are running VSCode with xwayland.  
Switching to wayland with `code --enable-features=UseOzonePlatform --ozone-platform=wayland` seems to finally fix my issues.  
No new setup of vscode required
